NewBank/newbank/server/NewBank.java

```diff
@@ -164,25 +177,147 @@ private String payRequest(CustomerID customerId, String[] splitRequest) {
+            return "FAIL";
+        }
+
+        // Convert second argument to double, return fail if not a number
+        double amount;
+        try {
+            amount = Double.parseDouble(splitRequest[1]);
+        } catch (NumberFormatException e) {
+            // amount in invalid format - not a number
+            return "FAIL";
+        }
+
+        // Check that amount requested is no more than the maximum, and is not a negative value
+        if (amount <= 0 || amount > 1000) {
+            return "FAIL";
+        }
+
+        ArrayList<Microloan> microloans = database.LoadMicroloans();
+
+        Customer customer = database.LoadCustomerReadWrite(customerId.getKey());
+        Account toAccount = customer.getAccount("Main");
+
+        // Eligibility check - Main account must have existed for at least 3 months
+        LocalDate mainOpeningDate = toAccount.getOpeningDate();
+        long diff = ChronoUnit.MONTHS.between(mainOpeningDate, LocalDate.now());
+        if (diff < 3) {
+            System.out.println("Main account must have existed for at least 3 months.");
+            database.SaveExistingCustomer(customer);
+            return "FAIL";
+        }
+
+        // Eligibility check - Cannot have more than 3 loans at a time
+        ArrayList<Microloan> customerMicroloans = database.GetCustomerMicroloans(customerId.getKey());
+        if (customerMicroloans.size() >= 3) {
+            System.out.println("Cannot have more than 3 active microloans at a time.");
+            database.SaveExistingCustomer(customer);
+            return "FAIL";
+        }
+
+        // Find active loaner with enough money in their Main account
+        String loanerId = this.database.findEligibleLoaner(amount);
+        if (loanerId == null) {
+            database.SaveExistingCustomer(customer);
+            return "FAIL";
+        }
+
+        Customer loaner = database.LoadCustomerReadWrite(loanerId);
+        Account fromAccount = loaner.getAccount("Main");
+
+        // Transfer money from loaner to requester
+        if (!fromAccount.removeMoney(amount)) {
+            database.SaveExistingCustomer(customer);
+            database.SaveExistingCustomer(loaner);
+            return "FAIL";
+        }
+        toAccount.addMoney(amount);
+
+        // Add microloan data to microloans file
+        int loanPeriod = 10;
+        double interestRate = 0.05;
+        microloans.add(new Microloan(loanerId, customerId.getKey(), LocalDateTime.now(), amount, loanPeriod, interestRate));
+
+        database.SaveMicroloans(microloans);
+
+        database.SaveExistingCustomer(customer);
+        database.SaveExistingCustomer(loaner);
+        return "SUCCESS";
+
+    }
+
+    /**
+     * This function should be executed periodically to pay back microloans which are due and
+     * update it to a historical database
+     */
+    public void updateMicroloans() {
+        System.out.println("Updating Microloans...");
```

Comment:
Do we want to remove this print statement too?

Reply:
I would leave this one, it just goes to the server output. But it's useful for observing when microloans are updated.